### PR TITLE
Improve linux toolchain script

### PR DESCRIPTION
**Pull request by networkfusion**
This will help with maintainability, and also with potential future support of Windows builds.

Current changes:
* Improve script header.
* Change dependency downloads to use HTTPS (rather than FTP).
* Seperate lines for all CMake Build options.
* Improve some comments.

Script has not (yet) been tested due to time required to actually build it.

### CODE REVIEW

File: tools/build [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
+# N64 MIPS GCC toolchain build/install script for Unix distributions
+# (c) 2012-2021 Shaun Taylor and libDragon Contributors.
+# See the root folder for license information.
 
-# You may be prompted for your password for installation steps.
-# If the script tells you that writing failed on chksum64 and n64tool,
-# this does not mean it failed.  The script simply retried with sudo
-# and your password was cached.
+
+### TODO: this info is probably incorrect and should be auto handled!
```
Comment (rasky) [inline tools/build]: I don't understand this TODO. The information looks correct to me. What you mean by "probably incorrect"?

File: tools/build [hunk 2]
```diff
@@ -18,8 +21,8 @@
 # Exit on error
 set -e
 
-# EDIT THIS LINE TO CHANGE YOUR INSTALL PATH!
-export INSTALL_PATH=${N64_INST:-/usr/local}
+# EDIT THIS LINE TO CHANGE YOUR INSTALL PATH! (TODO: !! Users should not be expected to do this !!)
+export INSTALL_PATH=${N64_INST:-/usr/local} #Perhaps this should be something like: "$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```
Comment (rasky) [inline tools/build]: The original comment is probably outdated. The way the `INSTALL_PATH` variable is defined, it takes the value of `N64_INST` if it exists. So users must not change this line: it's sufficient they define `N64_INST` before invoking the script. 

If you want, you can update the original comment just to say "Set N64_INST before calling the script to change the installation directory".

File: tools/build [hunk 3]
```diff
@@ -28,39 +31,45 @@ export FORCE_DEFAULT_GCC=${FORCE_DEFAULT_GCC:-false}
...
+# Dependency source: Extract stage
 test -d binutils-$BINUTILS_V || tar -xzf binutils-$BINUTILS_V.tar.gz
 test -d gcc-$GCC_V || tar -xzf gcc-$GCC_V.tar.gz
 test -d newlib-$NEWLIB_V || tar -xzf newlib-$NEWLIB_V.tar.gz
 
-# Use same native GCC
-
+# Check if the native system already has a pre-compiled compatible version of GCC installed.
 if [ ! $(gcc --version | grep $GCC_V) ] && [ $FORCE_DEFAULT_GCC == 'false' ]; then
+#TODO: have we tried installing it??? It might be faster!!
```
Comment (rasky) [inline tools/build]: I don't understand this TODO.

File: tools/build [hunk 4]
```diff
@@ -28,39 +31,45 @@ export FORCE_DEFAULT_GCC=${FORCE_DEFAULT_GCC:-false}
...
+#if not, build and install a compatible version of GCC for the native system.
```
Comment (rasky) [inline tools/build]: Please indent this.

File: tools/build [hunk 5]
```diff
@@ -28,39 +31,45 @@ export FORCE_DEFAULT_GCC=${FORCE_DEFAULT_GCC:-false}
...
     cd build_binutils
-    ../binutils-$BINUTILS_V/configure --prefix=${INSTALL_PATH}/gcc-$GCC_V --disable-werror
+    ../binutils-$BINUTILS_V/configure \
+      --prefix=${INSTALL_PATH}/gcc-$GCC_V \
+      --disable-werror
     make -j$JOBS
     make install || sudo make install || su -c "make install"
 
-    # Compile gcc (native)
+    # Compile gcc (native) system
```
Comment (rasky) [inline tools/build]: I think you want `(native system)`

File: tools/build [hunk 6]
```diff
@@ -76,30 +85,70 @@ mkdir gcc_compile
...
-# Compile gcc (pass 1)
+# Compile GCC for MIPS N64 (pass 1)
 cd ../gcc_compile
-../gcc-$GCC_V/configure --prefix=${INSTALL_PATH} --target=mips64-elf --with-arch=vr4300 --with-tune=vr4300 --enable-languages=c --without-headers --with-newlib --disable-libssp --enable-multilib --disable-shared --with-gcc --disable-threads --disable-win32-registry --disable-nls --disable-werror --with-system-zlib
+../gcc-$GCC_V/configure \
+  --prefix=${INSTALL_PATH} \
+  --target=mips64-elf \
+  --with-arch=vr4300 \
+  --with-tune=vr4300 \
+  --enable-languages=c \ #surely should include c++? If not, comment why!
```
Comment (rasky) [inline tools/build]: You don't need C++ for pass 1, as far as I know.

**Review by rasky — COMMENTED** [review]
I left a few comments. Also, really, you should test it at least once, otherwise we can't really commit it. Maybe you could add a Github action script here, so that it gets executed every time `tools/build` changes and rebuilds the toolchain.

File: tools/build (hunk 1 above)
Comment (networkfusion) [reply tools/build]: This build script is only for GCC, N64Tool and CheckSum are part of a different script are they not?

Comment (rasky) [reply tools/build]: Right. Just remove the comment then.

File: tools/build [hunk 7]
```diff
@@ -76,30 +79,70 @@ mkdir gcc_compile
...
-# Compile gcc (pass 1)
+# Compile GCC for MIPS N64 (pass 1)
 cd ../gcc_compile
-../gcc-$GCC_V/configure --prefix=${INSTALL_PATH} --target=mips64-elf --with-arch=vr4300 --with-tune=vr4300 --enable-languages=c --without-headers --with-newlib --disable-libssp --enable-multilib --disable-shared --with-gcc --disable-threads --disable-win32-registry --disable-nls --disable-werror --with-system-zlib
+../gcc-$GCC_V/configure \
+  --prefix=${INSTALL_PATH} \
+  --target=mips64-elf \
+  --with-arch=vr4300 \
+  --with-tune=vr4300 \
+  --enable-languages=c \ # c++ language support is added in the second pass.
```
Comment (rasky) [inline tools/build]: Unfortunately this doesn't work with bash. The continuation character `\` must be the last character on a line. You can't add the comment afterwards (you cannot even a space after it).

Comment (networkfusion) [reply tools/build]: Oops, I forgot about that. I will remove it. BTW, something in the back of my head tells me that C++ should also be enabled in the first pass, but I cannot remember why... But I guess it is easier to work out with the parameters on seperate lines if it ever becomes an issue.

**Comment by networkfusion** [comment]
> 
> 
> I left a few comments. Also, really, you should test it at least once, otherwise we can't really commit it. Maybe you could add a Github action script here, so that it gets executed every time `tools/build` changes and rebuilds the toolchain.

Just thinking about this... The best bet is for someone to just run this branch on a fresh unix install (with prerequsites met) and see if it performs the same (it should since the changes are all superficial)... Perhaps I can mention it on the n64brew discord... The automated build should be part of a seperate PR, and it might be worth autogenerating the docker container in that case, for which I am wondering if @anacierdem already has something...

**Comment by anacierdem** [comment]
I havent looked at this in detail yet (I will) but building the toolchain is _slow_. So it might not be possible with a free tier github action to release it to docker hub. I currently do it manually by generally separating the docker config [here](https://github.com/anacierdem/libdragon-docker/blob/master/Dockerfile) to layers in case there is a broken step. Then after making sure it works ok, I merge them to a single command and publish so that the size is small. Doing changes in the toolchain is generally time consuming if you intend to make it reproducible. Even without any change it can take about an hour on a 8 core 3GHz+ machine. It may thus hit the [time limit](https://github.com/anacierdem/libdragon-docker/blob/master/Dockerfile) of 6 hours if the processing power is low. We need to experiment for an automated build. I'm planning to move away from travis sometime so it might be a good time to try it on gh actions. We definitely cannot make this a PR check on libdragon. It _might_ make sense if it is only triggered by the build script though. Then it shouldnt have any code dependencies from the repo.

**Comment by networkfusion** [comment]
@anacierdem In that regard, it also might be worth checking out the PR here: https://github.com/rasky/libdragon/pull/6/files , but that will probably take some shared effort to make it truly usable.

File: tools/build (hunk 7 above)
Comment (networkfusion) [reply tools/build]: e.g. Would it fix: 
image - unlikely, but might be related.

Comment (rasky) [reply tools/build]: `__assert_func` is defined in `system.c` which is in `libdragonsys.a`. I see that it's being linked. So I don't know why the linker complains. That test compiles correctly for me, FWIW.

**Comment by rasky** [comment]
The PR is fine by me, but I guess it should be run at least one to make sure it works correctly.

File: tools/build (hunk 7 above)
Comment (networkfusion) [reply tools/build]: I have just merged in the latest changes from your `trunk`. The issue has resolved itself.

**Comment by rasky** [comment]
Thanks, we're including this in:https://github.com/DragonMinded/libdragon/pull/174

**Pull request closed without merging**